galera/src/replicator_str.cpp

```diff
@@ -105,6 +105,19 @@ ReplicatorSMM::sst_received(const wsrep_gtid_t& state_id,
     sst_cond_.signal();
 
 #ifdef PXC
+    // If the donor server crashed while SST is in progress, then SST script
```

Comment:
Hi Kamil!

Thank you very for your time on reviewing this PR.

Before answering your questions, let me give some background of the issue.

Without this fix, the current thread(SST joiner thread) never returns from this function as it waits forever on the `sst_mutex_` held by wsrep_replication_process thread, which shall be waiting for the start transfer to happen.
```
SST Joiner thread waiting for the lock which is held by the replication thread.

Thread 7 (Thread 0x7f1cc77fe700 (LWP 21403)):
#0  __lll_lock_wait (futex=futex@entry=0x8e109d0, private=0) at lowlevellock.c:52
#1  0x00007f1ce91c60a3 in __GI___pthread_mutex_lock (mutex=mutex@entry=0x8e109d0) at ../nptl/pthread_mutex_lock.c:80
#2  0x00007f1cd70b310a in gu::Mutex::lock (this=0x8e109d0) at galerautils/src/gu_mutex.hpp:55
#3  gu::lock::Lock (mtx=..., this=<synthetic pointer>) at galerautils/src/gu_lock.hpp:42
#4  galera::ReplicatorSMM::sst_received (this=0x8e10630, state_id=..., state=<optimized out>, rcode=-32) at galera/src/replicator_str.cpp:86
#5  0x00000000053c1ad0 in wsrep::wsrep_provider_v26::sst_received (this=0x8dd9500, gtid=..., err=-32) at /home/venki/work/pxc/3159/80/wsrep-lib/src/wsrep_provider_v26.cpp:947
#6  0x00000000053a75f7 in wsrep::server_state::sst_received (this=0x8dd91f0, cs=..., error=-32, awaiting_callback=0x7541c58 <sst_awaiting_callback>) at /home/venki/work/pxc/3159/80/wsrep-lib/src/server_state.cpp:775
#7  0x000000000333f4d4 in wsrep_sst_complete (thd=0x7f1cb0005350, rcode=-32) at /home/venki/work/pxc/3159/80/sql/wsrep_sst.cc:151
#8  0x000000000334527b in sst_joiner_thread (a=0x7f1ccc4e5748) at /home/venki/work/pxc/3159/80/sql/wsrep_sst.cc:609
#9  0x00007f1ce91c3609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#10 0x00007f1ce88f1103 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```
```
Wsrep replication thread waiting for the signal on completion of state transfer from the receiver thread.

Thread 6 (Thread 0x7f945aaf7700 (LWP 34626)):
#0  futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7f945aaf57e0) at ../sysdeps/nptl/futex-internal.h:183
#1  __pthread_cond_wait_common (abstime=0x0, clockid=0, mutex=0x7f945aaf5790, cond=0x7f945aaf57b8) at pthread_cond_wait.c:508
#2  __pthread_cond_wait (cond=cond@entry=0x7f945aaf57b8, mutex=mutex@entry=0x7f945aaf5790) at pthread_cond_wait.c:638
#3  0x00007f946623ea3d in gcs_replv (conn=conn@entry=0x8348220, act_in=act_in@entry=0x7f945aaf5820, act=act@entry=0x7f945aaf5890, scheduled=scheduled@entry=false) at gcs/src/gcs.cpp:1940
#4  0x00007f946623ecb9 in gcs_repl (conn=conn@entry=0x8348220, action=action@entry=0x7f945aaf5890, scheduled=scheduled@entry=false) at gcs/src/gcs.hpp:241
#5  0x00007f946623fc12 in gcs_request_state_transfer (conn=0x8348220, version=version@entry=3, req=req@entry=0x7f9444016870, size=size@entry=124, donor=<optimized out>, ist_gtid=..., order=@0x7f945aaf5b18: -1) at gcs/src/gcs.cpp:2056
#6  0x00007f9466291a15 in galera::Gcs::request_state_transfer (this=this@entry=0x833ceb0, version=3, req=0x7f9444016870, req_len=req_len@entry=124, sst_donor="", ist_gtid=..., seqno_l=@0x7f945aaf5b18: -1) at galera/src/galera_gcs.hpp:199
#7  0x00007f946629ecbc in galera::ReplicatorSMM::send_state_request (this=this@entry=0x833c6b0, req=req@entry=0x7f944401f010, unsafe=unsafe@entry=true) at galera/src/replicator_str.cpp:825
#8  0x00007f94662a1b62 in galera::ReplicatorSMM::request_state_transfer (this=this@entry=0x833c6b0, recv_ctx=recv_ctx@entry=0x7f945aaf6b58, group_uuid=..., cc_seqno=<optimized out>, cc_seqno@entry=2, sst_req=0x7f9444013a00, sst_req_len=47) at galera/src/replicator_str.cpp:1003
#9  0x00007f9466288ba1 in galera::ReplicatorSMM::process_conf_change (this=0x833c6b0, recv_ctx=0x7f945aaf6b58, cc=...) at galera/src/replicator_smm.cpp:2952
#10 0x00007f946626fa70 in galera::GcsActionSource::dispatch (this=this@entry=0x8348710, recv_ctx=recv_ctx@entry=0x7f945aaf6b58, act=..., exit_loop=@0x7f945aaf677e: false) at galera/src/gcs_action_source.cpp:124
#11 0x00007f946626fdef in galera::GcsActionSource::process (this=0x8348710, recv_ctx=0x7f945aaf6b58, exit_loop=@0x7f945aaf677e: false) at galera/src/gcs_action_source.cpp:182
#12 0x00007f946628730d in galera::ReplicatorSMM::async_recv (this=0x833c6b0, recv_ctx=0x7f945aaf6b58) at galera/src/replicator_smm.cpp:466
#13 0x00007f94662a7294 in galera_recv (gh=<optimized out>, recv_ctx=<optimized out>) at galera/src/wsrep_provider.cpp:236
#14 0x00000000053c0149 in wsrep::wsrep_provider_v26::run_applier (this=0x8305710, applier_ctx=0x7f945aaf6b58) at /home/venki/work/pxc/3159/80/wsrep-lib/src/wsrep_provider_v26.cpp:715
#15 0x000000000334ce4a in wsrep_replication_process (thd=0x7f9444000b80, arg=0x0) at /home/venki/work/pxc/3159/80/sql/wsrep_thd.cc:58
#16 0x0000000002e15fcb in start_wsrep_THD (arg=0x837d740) at /home/venki/work/pxc/3159/80/sql/mysqld.cc:9352
#17 0x0000000004ff72a8 in pfs_spawn_thread (arg=0x83658c0) at /home/venki/work/pxc/3159/80/storage/perfschema/pfs.cc:2855
#18 0x00007f9478420609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#19 0x00007f9477b4e103 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

Here, the replication thread will be waiting for the state transfer action to be delivered to it.
```
  1938     ┆   ┆   /* now we can go waiting for action delivery */
  1939             if (ret >= 0) {
  1940     ┆   ┆   ┆   gu_cond_wait (&repl_act.wait_cond, &repl_act.wait_mutex);
```

In normal cases, the abpve `repl_act.wait_cond` cond var will be signalled by the GCS recevier thread on receiving the state transfer from the donor. It signals from
```
  1542     ┆   ┆   gu_mutex_lock   (&repl_act->wait_mutex);
  1543     ┆   ┆   gu_cond_signal  (&repl_act->wait_cond);
  1544     ┆   ┆   gu_mutex_unlock (&repl_act->wait_mutex);
```
When donor goes away, the `gcs_recv_thread` shall never receive the state transfer and thus never signals the replication thread. Below is the stack trace of the receiver thread
```
GCS recevier thread waiting for an update on its queue.

Thread 4 (Thread 0x7f945b340700 (LWP 34624)):
#0  futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x837dc18) at ../sysdeps/nptl/futex-internal.h:183
#1  __pthread_cond_wait_common (abstime=0x0, clockid=0, mutex=0x837dbb8, cond=0x837dbf0) at pthread_cond_wait.c:508
#2  __pthread_cond_wait (cond=cond@entry=0x837dbf0, mutex=0x837dbb8) at pthread_cond_wait.c:638
#3  0x00007f946614ba62 in gu::Lock::wait (this=this@entry=0x7f945b33fa10, cond=...) at galerautils/src/gu_lock.hpp:78
#4  0x00007f9466244800 in RecvBuf::front (this=this@entry=0x837dbb8, timeout=...) at gcs/src/gcs_gcomm.cpp:137
#5  0x00007f9466242813 in gcomm_recv (backend=<optimized out>, msg=0x8348510, timeout=9223372035999999999) at gcs/src/gcs_gcomm.cpp:744
#6  0x00007f9466236aac in core_msg_recv (backend=backend@entry=0x8348670, recv_msg=recv_msg@entry=0x8348510, timeout=timeout@entry=9223372035999999999) at gcs/src/gcs_core.cpp:480
#7  0x00007f9466239d5c in gcs_core_recv (conn=0x83484b0, recv_act=recv_act@entry=0x7f945b33fd20, timeout=9223372035999999999) at gcs/src/gcs_core.cpp:1184
#8  0x00007f9466241023 in gcs_recv_thread (arg=0x8348220) at gcs/src/gcs.cpp:1471
#9  0x00007f9478420609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#10 0x00007f9477b4e103 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

Summarizing,
1. The receiver thread is inside `gcs_core_recv` waiting for update on the receiver queue,(which will never receive anything since donor is away).
2. Replication thread waiting for the `gcs_recv_thread` to receive the signal to complete the GCS_ACT_CCHANGE.
3. SST joiner thread waiting on the `sst_mutex_` held by the replication thread.

This link between the Joiner thread, replication thread and the receiver thread causes the joiner node to stall.

Answering your questions,
1. Without this fix, the  shall never go to wait on `sst_cond_` in `ReplicatorSMM::request_state_transfer()`. And thus `(sst_state_ == SST_CANCELED)` check wil also be not performed.
2. Regarding the parallel abort requests, because of the above deadlock, the replication thread never reaches the `cond_wait` and thus it cannot raise parallel abort requests.